File: hackerrank/problem solving/algorithms/2 - implementation/1 - grading students/grading_students.py
# https://www.hackerrank.com/challenges/grading/problem?isFullScreen=true

def main():
    n = int(input())
    grades = read_numbers(n)
    grading_students(grades)
    print(*grades, sep='\n')


def read_numbers(n):
    return [int(input()) for _ in range(n)]


# n: length of array grades
# T: O(n)
# S: O(1) extra space
def grading_students(grades):
    MIN_GRADE = 38

    for i, grade in enumerate(grades):
        if grade >= MIN_GRADE:
            next_multiple_5 = (grade // 5 + 1) * 5
            if next_multiple_5 - grade < 3:
                grades[i] = next_multiple_5


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/1 - grading students/grading_students.cpp
// https://www.hackerrank.com/challenges/grading/problem?isFullScreen=true
// C++23

import std;
using namespace std;

vector<int> read_numbers(int n);
vector<int>& grading_students(vector<int>& grades);

int main()
{
    int n;
    cin >> n;
    vector grades{ read_numbers(n) };
    for (auto x : grading_students(grades))
        println("{}", x);
    return 0;
}

vector<int> read_numbers(int n)
{
    vector<int> arr(n);
    for (auto& x : arr)
        cin >> x;
    return arr;
}

// n: length of grades
// T: O(n)
// S: O(1) extra space
vector<int>& grading_students(vector<int>& grades)
{
    for (constexpr int min_grade{ 38 }; auto& grade : grades) {
        if (grade >= min_grade) {
            int next_multiple_5{ (grade / 5 + 1) * 5 };
            if (next_multiple_5 - grade < 3)
                grade = next_multiple_5;
        }
    }
    return grades;
}